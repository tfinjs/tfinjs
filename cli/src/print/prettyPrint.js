import figures from 'figures';
import { join } from 'path';
import chalk from 'chalk';
import {
  getCyclic,
  getDependenciesOfNode,
  getInversedDependenciesOfNode,
  getNonCyclic,
} from '@tfinjs/dependency-graph';

const prettyPrint = ({
  add, graph, remove, update,
}, dist) => {
  const getDir = (d) => join(dist, d);
  let output = '';
  const cyclic = getCyclic(graph);
  if (cyclic.stacks.length > 0) {
    const content = cyclic.stacks.map(
      (nodes) => `${figures.bullet} ${nodes.join(', ')}`,
    );
    output += `
      ${chalk.red.bold('Warning, you have circular dependencies!')}\n
      ${chalk.red(content)}
    `;
  }

  const dependencies = getNonCyclic(graph);
  const levels = dependencies
    .map((nodes) => nodes.filter((node) => [...add, ...update].includes(node)))
    .filter((nodes) => nodes.length);
  if (levels.length) {
    output += `\n\n${chalk.green.bold('Run the following to deploy')}
    ${levels
    .map(
      (nodes, index) =>
        `\n${chalk(`# level ${index}:`)}\n${nodes
          .map((node) =>
            chalk.bgBlack.green(
              `\ncd ${chalk.bold(
                getDir(node),
              )} && tf init && tf apply -auto-approve`,
            ))
          .join(' &&\n')}`,
    )
    .join('\n\n')}
    `;
  } else {
    output += '\nNothing to deploy\n';
  }
  if (remove.length) {
    output += `\n\n${chalk.red.bold('Delete the following resources')}
    ${chalk.bgBlack.red(
    remove
      .filter((node) => remove.includes(node))
      .map(
        (node) =>
          `\ncd ${chalk.bold(
            getDir(node),
          )} && tf init && tf destroy -auto-approve`,
      )
      .join(' &&\n'),
  )}
  `;
  }

  return output;
};

export default prettyPrint;
