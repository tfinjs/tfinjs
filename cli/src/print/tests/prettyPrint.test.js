import prettyPrint from '../prettyPrint';

/* eslint-env jest */


test('prettyPrint cyclic', () => {
  const output = prettyPrint(
    {
      add: ['tijpetshop1lr7f2m'],
      graph: { tijpetshop1lr7f2m: ['tijpetshop1lr7f2m'], tijpetshopmjubxt: [] },
      remove: [],
      update: [],
    },
    '/',
  );
  expect(output).toMatchSnapshot();
});

test('prettyPrint normal', () => {
  const output = prettyPrint(
    {
      add: ['tijpetshop1lr7f2m'],
      graph: { tijpetshop1lr7f2m: [], tijpetshopmjubxt: [] },
      remove: [],
      update: [],
    },
    '/',
  );
  expect(output).toMatchSnapshot();
});
test('prettyPrint remove', () => {
  const output = prettyPrint(
    {
      add: ['tijpetshop1lr7f2m'],
      graph: { tijpetshop1lr7f2m: [], tijpetshopmjubxt: [] },
      remove: ['wefwef', 'fewfew'],
      update: [],
    },
    '/',
  );
  expect(output).toMatchSnapshot();
});

test('prettyPrint normal with graph', () => {
  const output = prettyPrint(
    {
      add: ['a', 'b', 'c', 'd'],
      graph: {
        a: ['b'],
        b: ['c', 'd'],
        c: [],
        d: [],
        e: [],
      },
      remove: [],
      update: [],
    },
    '/',
  );
  expect(output).toMatchSnapshot();
});
