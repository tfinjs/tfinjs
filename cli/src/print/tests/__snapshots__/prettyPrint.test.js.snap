// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettyPrint cyclic 1`] = `
"
      [31m[1mWarning, you have circular dependencies![22m[39m

      [31m‚óè tijpetshop1lr7f2m[39m
    
Nothing to deploy
"
`;

exports[`prettyPrint normal 1`] = `
"

[32m[1mRun the following to deploy[22m[39m
    
# level 0:
[40m[32m[39m[49m
[40m[32mcd [1m/tijpetshop1lr7f2m[22m && tf init && tf apply -auto-approve[39m[49m
    "
`;

exports[`prettyPrint normal with graph 1`] = `
"

[32m[1mRun the following to deploy[22m[39m
    
# level 0:
[40m[32m[39m[49m
[40m[32mcd [1m/c[22m && tf init && tf apply -auto-approve[39m[49m &&
[40m[32m[39m[49m
[40m[32mcd [1m/d[22m && tf init && tf apply -auto-approve[39m[49m


# level 1:
[40m[32m[39m[49m
[40m[32mcd [1m/b[22m && tf init && tf apply -auto-approve[39m[49m


# level 2:
[40m[32m[39m[49m
[40m[32mcd [1m/a[22m && tf init && tf apply -auto-approve[39m[49m
    "
`;

exports[`prettyPrint remove 1`] = `
"

[32m[1mRun the following to deploy[22m[39m
    
# level 0:
[40m[32m[39m[49m
[40m[32mcd [1m/tijpetshop1lr7f2m[22m && tf init && tf apply -auto-approve[39m[49m
    

[31m[1mDelete the following resources[22m[39m
    [40m[31m[39m[49m
[40m[31mcd [1m/wefwef[22m && tf init && tf destroy -auto-approve &&[39m[49m
[40m[31m[39m[49m
[40m[31mcd [1m/fewfew[22m && tf init && tf destroy -auto-approve[39m[49m
  "
`;
